[FEA] make raft sources compilable with clang

MatthiasKohl opened the pull request:
This makes RAFT sources compilable with clang.
It fixes some fragile code (using `static const` instead of `static constexpr` or `%laneid` in PTX relying on quirks in nvcc which make this happen).

RAFT is still not compilable with clang entirely though due to the dependencies:
1. cub has this issue before 1.14: https://github.com/NVIDIA/cub/pull/335
2. libcudacxx has issues with atomic, which should be fixed in >= 1.7.0-ea (wasn't able to verify this yet)
3. libcudacxx has issues with variadic CUDA functions, which is apparently fixed by passing `-Xclang -fcuda-allow-variadic-functions` to clang (wasn't able to verify this yet)
3. cooperative_groups from CUDA does not work with clang 11.0 / 11.1 but only with >= 13

EDIT: this is necessary to close #84

MatthiasKohl commented:
question to reviewers: what would be the easiest way to bump up versions of tools and dependencies ?
In particular, to make this really happen, we would need thrust/cub >= 1.14, libcudacxx >= 1.7.0-ea and clang-tools >= 13.

With the Rapids cpm system, this seems prohibitively hard. Why?

cjnolet commented:
@MatthiasKohl, I know you've put this down in the meantime due to some issues w/ dependent versions in RAPIDS. I still see value in merging the c++ code changes and keeping the script in a PR until clang support is ready to go. What do you think?

MatthiasKohl commented:
Yes I think it makes sense to add these changes to RAFT. But it's hard to test because of these issues in dependencies meaning that for many files, clang doesn't even get to the RAFT part. In fact after I upgraded to clang 13 and updated some of the dependent packages, I still discovered new issues within RAFT sources because they were simply not compiled before.
So what we could do is to just ignore the errors for now and gradually change things as more and more code will be covered with newer packages and newer clang versions

MatthiasKohl commented:
@cjnolet I've added a few changes I had in the pipeline which make things compatible with CUDA 11.5, plus a few other RAFT sources which can only be compiled with a newer clang version and updated thrust.

Can you review these changes so that we can already merge those?

Right now, libcudacxx and cuco cannot be compiled device-side.
I can still check whether they can be compiled host-side which would already be nice.
But I feel like that isn't really enough for a project like RAFT where probably 50% of code is device-side, and especially the important code is device-side, so before spending more time on this, I'd prefer having the dependent projects work with device-side compilation.

MatthiasKohl commented:
rerun tests

cjnolet commented on cpp/scripts/run-clang-compile.py:
```diff
@@ -0,0 +1,320 @@
+# Copyright (c) 2020-2021, NVIDIA CORPORATION.
```
Just a reminder to update copyrights.

MatthiasKohl replied:
I've found more than 300 files in RAFT not passing the copyright checker, I'll probably file a separate PR for this, although I'll update the copyrights of the files in this particular PR as well.

cjnolet commented on cpp/scripts/run-clang-compile.py:
```diff
@@ -0,0 +1,320 @@
+import subprocess
+
+
+CLANG_COMPILER = "clang++"
```
Just to clarify current status: we won't expect this compile successfully yet, correct? If not, we should add a note to this file, maybe underneath the license, which states clang support is experimental/not fully working yet. In that case, I'm also okay getting the other changes in this PR merged but pushing this script into another PR.

MatthiasKohl replied:
Yes exactly, we don't expect successful compilation, at least for some files in the RAFT test suite. I'll add a note there

cjnolet commented on cpp/include/raft/cuda_utils.cuh:
```diff
@@ -109,7 +109,7 @@ static const int WarpSize = 32;
 {
   int id;
-  asm("mov.s32 %0, %laneid;" : "=r"(id));
+  asm("mov.s32 %0, %%laneid;" : "=r"(id));
```
Is there some sort of documentation or guidelines we can add to our new [developer guide](https://github.com/rapidsai/raft/blob/branch-22.02/DEVELOPER_GUIDE.md) which would make it easier for developers to follow these rules in the future?

MatthiasKohl replied:
It's part of the inline PTX assembly spec: https://docs.nvidia.com/cuda/inline-ptx-assembly/index.html#parameters (quote : "If you want the `%` in a ptx instruction, then you should escape it with double `%%`, e.g.: [...] ").
I honestly have no idea why `nvcc` even compiles the above code, I guess they have added some special cases for the special registers against their own docs.
I can add a link to the spec in the developer guide, but IMO, this is more confusing than anything. The underlying issue is that `nvcc` accepts this (clearly incorrect) code.

cjnolet reviewed (changes requested):
Since RAFT is still somewhat coupled to cuml's C++ code, can you open a PR in cuml w/ the `get_raft.cmake` pinned to this branch just so we can make sure it doesn't break? Overall the changes look good.

MatthiasKohl commented:
@cjnolet sorry I didn't see your comments on this PR (somehow missed the notifications), I'll address them ASAP

MatthiasKohl commented:
@cjnolet you might have seen this draft PR on cuml, which I'm using to make sure that cuml CI is still passing with these changes: https://github.com/rapidsai/cuml/pull/4571

Note that I had to add some more changes to raft in order to make sure that RAFT sources themselves are compiling with Clang: - [cpp/include/raft/cuda_utils.cuh](https://github.com/rapidsai/raft/pull/424/files#diff-18c5a633557f949d60d48193bb42d9adf38c34f4454594bcec6552e7ece60428) lines 697-728 : I added overloads for the `__ldcv/__ldcg/__stwt` CUDA functions which are unknown to clang
- https://github.com/rapidsai/raft/pull/424/files#diff-ca87b707f7d1c64a1085887c53fc4e80799a0b82ccc19a4aff3471141ad79dbb lines 317-320 : I excluded the `__shared__ alignas` construct from clang compilation since Clang cannot compile that unfortunately.

MatthiasKohl commented:
@cjnolet CI on cuml is passing with the latest changes from this PR (see https://github.com/rapidsai/cuml/pull/4571)
It seems like CI here on the RAFT PR isn't being run for some reason.

Apart from that, I'm still awaiting your final review for the latest changes since the compiler-specific stuff might not be what you'd like to do for RAFT in general.

cjnolet commented:
rerun tests

cjnolet reviewed (changes requested):
I'm still thinking through the best way to isolate and document the files that can be safely compiled with a CUDA vs non-CUDA-enabled compiler while keeping the API consistent. 

Thanks again for these changes! I really just have a couple minor-ish questions and one small change.

achirkin commented on cpp/include/raft/matrix/detail/linewise_op.cuh:
```diff
@@ -312,8 +314,12 @@ __global__ void __launch_bounds__(BlockSize)
   typedef Linewise<Type, IdxType, VecBytes, BlockSize> L;
   constexpr uint workSize = L::VecElems * BlockSize;
   uint workOffset         = workSize;
+#if defined(__clang__)
```
Yes, I'd prefer to pull this out as well. Perhaps, into something like `RAFT_alignas`, since it may be useful in cuml too?

MatthiasKohl replied:
This one is the most tricky since clang doesn't know about this construct at all.
We stumbled upon a similar thing in cudf where clang doesn't allow `__device__ [[nodiscard]] ... method(...)` in classes, even though without the `__device__` it would be perfectly fine.
Do you think it would be acceptable to use `__align(n)__` instead?
I'm not sure if in CUDA that is precisely the same as `alignas` in C++, but it would solve the issue with clang at least.

achirkin commented on cpp/include/raft/cuda_utils.cuh:
```diff
@@ -694,6 +694,41 @@ DI T blockReduce(T val, char* smem)
+__device__ __forceinline__ void stwt(T* addr, const T& val)
+{
+#if defined(__clang__)
+  *addr = val;
```
Maybe we'd need some `volatile`-casting tricks here to make clang backend avoid caches as well?

MatthiasKohl commented on cpp/include/raft/cluster/detail/kmeans.cuh:
```diff
@@ -324,6 +325,16 @@ static __global__ void divideCentroids(index_type_t d,
+ * Ceiling division operation
+ * Re-declared here from raft/cuda_utils.cuh since we cannot include CUDA files
+ */
+template <typename IntType>
```
Thank you for raising this again!
Given the comment, I must have thought that this file was compiled from a `.cpp` file (thus with a non-CUDA-enabled compiler). I actually don't think this was the case, and these changes don't make any sense to me anymore, so I reverted them and they don't seem to affect clang compilation at all.

MatthiasKohl commented on cpp/include/raft/cuda_utils.cuh:
```diff
@@ -694,6 +694,41 @@ DI T blockReduce(T val, char* smem)
+template <typename T>
+__device__ __forceinline__ T ldcg(const T* addr)
+{
+#if defined(__clang__)
```
I think I found a better way of doing this, I'll revert this change.
Thanks for raising it and adding @achirkin !

MatthiasKohl replied:
Basically, I believe what we should do here is to pre-define these functions for clang when compiling with it, and only when compiling with clang, just like clang defines the other CUDA intrinsics / global device functions (https://github.com/llvm-mirror/clang/blob/master/lib/Headers/__clang_cuda_device_functions.h).
@achirkin : given your other comment, for now, I would not actually try to ensure that clang produces anything competitive in terms of a binary. However, being able to parse the syntax of RAFT with clang allows tool like `clang-tidy` to be used, and this is really what I wanted to achieve with this PR.

achirkin replied:
Ok, then, maybe I would add a comment for future - if someone wants to make clang compiler generate correct kernel code w.r.t. memory fences.

MatthiasKohl replied:
Yeah, this makes sense, I'll play around with it, maybe we can use the assembly instructions directly for clang.
In any case, we won't really emit the same code with clang because optimizations are different etc.
But we can come close to it probably.

MatthiasKohl commented:
@cjnolet @achirkin would it be possible to review the latest changes again? I've made the following changes:
1. Using `__align__` instead of `alignas` in device code which clang accepts
2. No more weird things around `__ldcv` and similar functions. I simply added these intrinsics in an include which is added for CUDA compilation with clang, see the new header file under `cpp/scripts`
3. I reverted the changes around `ceildiv` in `cluster/detail/kmeans.cuh`

cjnolet reviewed (approved):
This looks good from my end. Thanks for doing this! It will be very useful once we are able to run clang-tidy across the entire codebase.

cjnolet commented:
Also hoping to get final thoughts from @achirkin before merging.

achirkin commented:
Looks great to me, looking forward to using `clang-tidy` in all cuda files!

achirkin reviewed (approved):
Looks great to me, looking forward to using `clang-tidy` in all cuda files!

cjnolet commented:
@gpucibot merge
